Implement AmberScript COPY command

hevrard opened the pull request:
I am not sure about the proper implementation in executor, in particular, how should we check that the destination buffer is clean? how should we properly clean it?

dj2 commented on src/amberscript/parser.cc:
```diff
@@ -60,14 +60,18 @@ Result Parser::Parse(const std::string& data) {
       r = ParseRun();
     } else if (tok == "SHADER") {
       r = ParseShaderBlock();
+    } else if (tok == "CLEAR") {
```
I think this was a bad merge, clear was moved up to line 53 to put these in sorted order (can you move COPY up there as well?)

dj2 commented on src/executor.cc:
```diff
@@ -111,6 +111,14 @@ Result Executor::Execute(Engine* engine,
+        copy->GetBufferTo()->SetMemPtr(nullptr);
+      }
+      std::vector<Value> data = copy->GetBufferFrom()->GetData();
+      copy->GetBufferTo()->SetData(std::move(data));
```
I think we want to put this method into buffer itself, then executor just does `Result r = buffer_from->CopyTo(buffer_to);`

Then in the buffer code you need to do something like:

````
// TODO(hevrard) Return error if buffer_to mem_ptr_ is backed by GPU memory
free(buffer_to->mem_ptr_);
buffer_to_->values_.clear();
bufer_to_values_ = values_;
if (mem_ptr_ != nullptr) {
  malloc(buffer_to->mem_ptr_, width * height * format byte size)
  std::memcpy(buffer_to->mem_ptr_, ....);
}
return {}
```

The GetMemPtr is a bit tricky as it returns either the mem_ptr_ or the values data buffer, and so we can't use it directly to do the copy.

This also means you'll need to be sure we're cleaning up mem_ptr_ correctly in this case. I think I assume'd it would get cleaned up through the GPU code?

hevrard replied:
Yeah I spotted that GetMemPtr is a bit tricky. Thanks for the guidance here.

So a buffer stores things in either data_, values_, or mem_ptr_? Is this exclusive, i.e., only one of these three is allowed at a time?

dj2 replied:
To be on the safe side, I'd just copy all of them. In general only one should be set, but if we don't need to make that assumption, I'd prefer not too.

hevrard commented on src/amberscript/parser.cc:
```diff
@@ -1115,5 +1119,57 @@ Result Parser::ParseExpect() {
+  if (buffer_to == nullptr) {
+    return Result("COPY destination buffer was not declared");
+  }
+
```
Good point!

hevrard commented on src/command.h:
```diff
@@ -194,6 +198,19 @@ class ComputeCommand : public Command {
 
+class CopyCommand : public Command {
+ public:
+  explicit CopyCommand(Buffer* buffer_from, Buffer* buffer_to);
```
It seems like all command constructors have an explicit keyword? e.g.

`explicit ProbeCommand(Pipeline* pipeline, Buffer* buffer);`

dj2 replied:
Yea, they got changed to include the pipeline and I forgot to remove the explicit. It doesn't hurt, it just doesn't do anything.

hevrard commented:
Now the copy properly checks that buffer dimensions are compatible, and copy relevant fields. I still need to check whether mem_ptr_ is properly cleaned in GPU code.

As you will notice, it looks like we have to cast to Data/Format buffers in order to get the relevant size in bytes for each of them.

More generally, the tests/case/copy.amber script used to fail with:
```
36: cannot COPY between buffers of different types
```
Because the destination is of type "unknown" since it is not used in the pipeline. The parser now sets the destination buffer type, if it is unknown, to match the origin buffer type.

Now the test fails with:
```
37: buffer not in pipeline for EXPECT command
```

The EXPECT logic seems to limit itself to buffers used in a pipeline, shall we extend this to buffers used as COPY destinations?

hevrard commented:
OK just saw #377, which should remove the current EXPECT error I guess.

dj2 commented:
Yea, the EXPECT pipeline requirement was needed to make it work, but has now been removed as I refactored the code. If you update on top of master hopefully that error should go away.

dj2 commented on src/buffer.cc:
```diff
@@ -30,16 +33,104 @@ FormatBuffer* Buffer::AsFormatBuffer() {
+
+  if (buffer->mem_ptr_) {
+    // TODO(hevrard) Return error if buffer_to mem_ptr_ is backed by GPU memory
+    std::free(buffer->mem_ptr_);
```
You can remove the if wrapper, it's safe to call free on a nullptr.

hevrard replied:
OK, I wasn't sure what is the best practice for Amber.

dj2 commented on src/buffer.cc:
```diff
@@ -30,16 +33,104 @@ FormatBuffer* Buffer::AsFormatBuffer() {
+  if (data_buffer->GetSizeInBytes() != GetSizeInBytes())
+    return Result("DataBuffer::CopyTo() buffers have a different byte size");
+
+  if (data_buffer->mem_ptr_) {
```
ditto

dj2 replied:
Actually, you could add a help in Buffer which does the mem_ptr_ copy. Just send in the size from the overridden CopyTo method. Removes the duplicated code.

hevrard replied:
Good point!

dj2 commented on src/buffer.cc:
```diff
@@ -30,16 +33,104 @@ FormatBuffer* Buffer::AsFormatBuffer() {
+  if (format_buffer->GetSizeInBytes() != GetSizeInBytes())
+    return Result("FormatBuffer::CopyTo() buffers have a different byte size");
+
+  if (format_buffer->mem_ptr_) {
```
ditto

dj2 commented on src/executor.cc:
```diff
@@ -111,6 +111,17 @@ Result Executor::Execute(Engine* engine,
+      auto copy = cmd->AsCopy();
+      auto buffer_from = copy->GetBufferFrom();
+      auto buffer_to = copy->GetBufferTo();
+      if (buffer_from->IsDataBuffer() && buffer_to->IsDataBuffer()) {
```
I don't think you need to do this. C++ should pick the overridden method by default so just doing `buffer_from->CopyTo(buffer_to)` should be fine I believe.

hevrard replied:
Thanks, and apologies for all these newbie bumps, I managed to stick to C so far in my career so I am a bit walking on eggs when dealing with C++ inheritance and overriding.

hevrard commented:
Thanks again for the feedback. I applied comments, rebased on master and added documentation.

The `copy_format_buffer.amber` test now fails because RUN is not implemented, I guess. I tried to use data buffers in `copy_data_buffer.amber`, but the EXPECT implementation is lacking.

`mem_ptr_` seems to be set only in:
```
src/vulkan/frame_buffer.cc:80:      info->buffer->SetMemPtr(color_images_.back()->HostAccessibleMemoryPtr());
```
In general, Buffer objects seems to be used only in Pipeline inside a BufferInfo struct. As there does not appear to exist any method that remove a buffer from the pipeline, I guess we cannot have dandling pointers in Buffer.mem_ptr_. I am not sure how to be able to check thoroughly whether the mem_ptr_ points to GPU memory?

dj2 commented:
PR #383 fixes the format copy command. I'm working on adding support for the EQ command in EXPECT calls and I'm working to remove the data_ pointer in the data buffers which should fixup your data format test.

Once #383 lands, I think you can rebase this on top of that and verify your copy format test works. Then rename the copy data test be .expect_fail.amber and then I can rename it to just .amber when I get the EQ and data CLs landed.

dj2 commented:
I have a CL up which deletes the data_ pointer from buffer and a second one ready which deletes the mem_ptr_ from buffer. Once they land, you just have to copy the values_ vector.

dj2 commented:
Ok, everything has landed and you should be able to simplify this to just copy the values_ vector in the base buffer class. Overrides aren't needed anymore.

hevrard commented:
Thanks for this! I'm busy on other fronts but will take care of this tomorrow.

dj2 commented on src/buffer.h:
```diff
@@ -93,6 +93,9 @@ class Buffer {
   }
 
+  /// Copy the buffer values to an other one
+  virtual Result CopyTo(Buffer* buffer) const;
```
Nit: remove virtual as we don't override this anymore

dj2 commented on src/buffer.cc:
```diff
@@ -210,6 +210,18 @@ FormatBuffer* Buffer::AsFormatBuffer() {
+    return Result("Buffer::CopyBaseFields() buffers have a different height");
+  if (buffer->size_ != size_)
+    return Result("Buffer::CopyBaseFields() buffers have a different size");
+  buffer->values_.clear();
```
nit: clear() isn't needed as the buffer will get removed when we assign it below.
